internal/install/recipes/recipe_detector.go

```diff
@@ -10,52 +10,94 @@ import (
+	Recipe     *types.OpenInstallationRecipe
+	Status     execution.RecipeStatusType
+	DurationMs int64
+}
+
 type DetectionStatusProvider interface {
 	DetectionStatus(context.Context, *types.OpenInstallationRecipe) execution.RecipeStatusType
 }
 
 type RecipeDetector struct {
-	processEvaluator DetectionStatusProvider
-	scriptEvaluator  DetectionStatusProvider
-	recipeEvaluated  map[string]bool // same recipe(ref) should only be evaluated one time
+	processEvaluator      DetectionStatusProvider
+	scriptEvaluator       DetectionStatusProvider
+	context               context.Context
+	repo                  Finder
+	recipeDetectionResult map[string]*RecipeDetectionResult
+	availableRecipes      map[string]*RecipeDetectionResult
+	unavaliableRecipes    map[string]*RecipeDetectionResult
+	isDetectionRunned     bool
 }
 
-func NewRecipeDetector() *RecipeDetector {
+func NewRecipeDetector(contex context.Context, repo *RecipeRepository) *RecipeDetector {
 	return &RecipeDetector{
-		processEvaluator: NewProcessEvaluator(),
-		scriptEvaluator:  NewScriptEvaluator(),
-		recipeEvaluated:  make(map[string]bool),
+		processEvaluator:      NewProcessEvaluator(),
+		scriptEvaluator:       NewScriptEvaluator(),
+		context:               contex,
+		repo:                  repo,
+		recipeDetectionResult: make(map[string]*RecipeDetectionResult),
+		availableRecipes:      make(map[string]*RecipeDetectionResult),
+		unavaliableRecipes:    make(map[string]*RecipeDetectionResult),
 	}
 }
 
-func (dt *RecipeDetector) detectBundleRecipe(ctx context.Context, bundleRecipe *BundleRecipe) {
-	// if already evaluated
-	if _, exists := dt.recipeEvaluated[bundleRecipe.Recipe.Name]; exists {
-		return
+func (dt *RecipeDetector) detectRecipes() error {
+	if dt.isDetectionRunned {
+		return nil
+	}
+
+	dt.isDetectionRunned = true
+
+	recipes, err := dt.repo.FindAll()
+	if err != nil {
+		return err
+	}
+
+	for _, r := range recipes {
+		dr := dt.detectRecipe(r)
+		dt.recipeDetectionResult[dr.Recipe.Name] = dr
+
+		if dr.Status == execution.RecipeStatusTypes.AVAILABLE {
+			dt.availableRecipes[dr.Recipe.Name] = dr
+		} else {
+			dt.unavaliableRecipes[dr.Recipe.Name] = dr
+		}
 	}
-	dt.recipeEvaluated[bundleRecipe.Recipe.Name] = true
+	return nil
+}
 
-	for _, dep := range bundleRecipe.Dependencies {
-		dt.detectBundleRecipe(ctx, dep)
+func (dt *RecipeDetector) GetAvaliableRecipes() (map[string]*RecipeDetectionResult, error) {
+	err := dt.detectRecipes()
+	if err != nil {
+		return nil, err
 	}
+	return dt.availableRecipes, nil
+}
 
-	if bundleRecipe.AreAllDependenciesAvailable() {
-		status, durationMs := dt.detectRecipe(ctx, bundleRecipe.Recipe)
-		bundleRecipe.AddDetectionStatus(status, durationMs)
-	} else {
-		log.Debugf("Dependency not available for recipe %v", bundleRecipe.Recipe.Name)
+func (dt *RecipeDetector) GetUnavaliableRecipes() (map[string]*RecipeDetectionResult, error) {
+	err := dt.detectRecipes()
+	if err != nil {
+		return nil, err
```

Comment:
@NRhzhao: Seems like these are both called only once in `recipe_installer` - maybe this could be `GetDetectedRecipes` and returns two maps - available and unavailable - since these are set in `detectRecipes`.  Then we could get rid of the `isDetectionRunned` concept

Reply:
@mbruzina Yeah, I was thinking about that, but opt to went with separate calls since it's simpler.  But I can update to your recommendation.